crates/astria-sequencer/src/accounts/action.rs

```diff
@@ -25,56 +26,74 @@ use crate::{
 /// Fee charged for a `Transfer` action.
 pub(crate) const TRANSFER_FEE: u128 = 12;
 
+pub(crate) async fn transfer_check_stateful<S: StateReadExt + 'static>(
+    action: &TransferAction,
+    state: &S,
+    from: Address,
+) -> Result<()> {
+    ensure!(
+        state.is_allowed_fee_asset(action.fee_asset_id).await?,
+        "invalid fee asset",
+    );
+
+    let transfer_asset_id = action.asset_id;
+
+    let from_fee_balance = state
+        .get_account_balance(from, action.fee_asset_id)
+        .await
+        .context("failed getting `from` account balance for fee payment")?;
+
+    // if fee asset is same as transfer asset, ensure accounts has enough funds
+    // to cover both the fee and the amount transferred
+    if action.fee_asset_id == transfer_asset_id {
+        let payment_amount = action
+            .amount
+            .checked_add(TRANSFER_FEE)
+            .ok_or(anyhow!("transfer amount plus fee overflowed"))?;
+
+        ensure!(
+            from_fee_balance >= payment_amount,
+            "insufficient funds for transfer and fee payment"
+        );
+    } else {
+        // otherwise, check the fee asset account has enough to cover the fees,
+        // and the transfer asset account has enough to cover the transfer
+        ensure!(
+            from_fee_balance >= TRANSFER_FEE,
+            "insufficient funds for fee payment"
+        );
+
+        let from_transfer_balance = state
```

Comment:
Nit: the naming here tripped me up. What about calling this just `from_balance`?

Reply:
there's two from balances though, this is to distinguish it from the `from_fee_balance`